File: test_sample.cpp
#include <iostream>
#include <cmath>
#include <limits>
#include <string>

/**
 * @brief This module provides basic arithmetic operations: addition, subtraction, 
 * multiplication, and division, with a check for division by zero.
 */

/**
 * @brief Adds two numbers.
 * * @param a The first number.
 * @param b The second number.
 * @return The sum of a and b.
 */
double add(double a, double b) {
    return a + b;
}

/**
 * @brief Subtracts the second number from the first.
 * * @param a The number to subtract from (minuend).
 * @param b The number to subtract (subtrahend).
 * @return The difference between a and b.
 */
double subtract(double a, double b) {
    return a - b;
}

/**
 * @brief Multiplies two numbers.
 * * @param a The first number.
 * @param b The second number.
 * @return The product of a and b.
 */
double multiply(double a, double b) {
    return a * b;
}

/**
 * @brief Divides the first number by the second. 
 * Includes a check for division by zero.
 * * @param a The dividend.
 * @param b The divisor.
 * @return The quotient of a and b. Returns NaN (Not a Number) if division by zero occurs.
 */
double divide(double a, double b) {
    if (b == 0.0) {
        // Return Not-a-Number (NaN) to signal an error.
        return std::numeric_limits<double>::quiet_NaN();
    }
    return a / b;
}

/**
 * @brief The main function demonstrates the usage of the arithmetic functions.
 */
int main() {
    std::cout << "--- Basic Arithmetic Demonstrator ---" << std::endl;

    // Example numbers
    double x = 10.0;
    double y = 5.0;
    double z = 0.0;

    // 1. Addition
    double result_add = add(x, y);
    std::cout << "\nAddition: " << x << " + " << y << " = " << result_add << std::endl;

    // 2. Subtraction
    double result_subtract = subtract(x, y);
    std::cout << "Subtraction: " << x << " - " << y << " = " << result_subtract << std::endl;

    // 3. Multiplication
    double result_multiply = multiply(x, y);
    std::cout << "Multiplication: " << x << " * " << y << " = " << result_multiply << std::endl;

    // 4. Division (Normal Case)
    double result_divide_ok = divide(x, y);
    // Check if the result is NaN (error condition)
    if (std::isnan(result_divide_ok)) {
        // This line is technically unreachable for x=10, y=5, but is included for robustness
        std::cout << "Division (OK): " << x << " / " << y << " = " << "Error: Cannot divide by zero." << std::endl;
    } else {
        std::cout << "Division (OK): " << x << " / " << y << " = " << result_divide_ok << std::endl;
    }

    // 5. Division (Division by Zero Check)
    double result_divide_error = divide(x, z);
    // Check if the result is NaN (error condition)
    if (std::isnan(result_divide_error)) {
        // We replicate the Python's explicit error message here
        std::cout << "Division (Error): " << x << " / " << z << " = " << "Error: Cannot divide by zero." << std::endl;
    } else {
        std::cout << "Division (Error): " << x << " / " << z << " = " << result_divide_error << std::endl;
    }

    return 0;
}

File: test_sample.py
"""
This module provides basic arithmetic operations: addition, subtraction, 
multiplication, and division, with a check for division by zero.
"""

def add(a: float, b: float) -> float:
    """
    Adds two numbers.

    Args:
        a: The first number.
        b: The second number.

    Returns:
        The sum of a and b.
    """
    return a + b

def subtract(a: float, b: float) -> float:
    """
    Subtracts the second number from the first.

    Args:
        a: The number to subtract from (minuend).
        b: The number to subtract (subtrahend).

    Returns:
        The difference between a and b.
    """
    return a - b

def multiply(a: float, b: float) -> float:
    """
    Multiplies two numbers.

    Args:
        a: The first number.
        b: The second number.

    Returns:
        The product of a and b.
    """
    return a * b

def divide(a: float, b: float) -> float or str:
    """
    Divides the first number by the second. Includes a check for division by zero.

    Args:
        a: The dividend.
        b: The divisor.

    Returns:
        The quotient of a and b, or a string error message if division by zero occurs.
    """
    if b == 0:
        return "Error: Cannot divide by zero."
    return a / b

def main():
    """
    The main function demonstrates the usage of the arithmetic functions.
    """
    print("--- Basic Arithmetic Demonstrator ---")

    # Example numbers
    x = 10
    y = 5
    z = 0

    # 1. Addition
    result_add = add(x, y)
    print(f"\nAddition: {x} + {y} = {result_add}")

    # 2. Subtraction
    result_subtract = subtract(x, y)
    print(f"Subtraction: {x} - {y} = {result_subtract}")

    # 3. Multiplication
    result_multiply = multiply(x, y)
    print(f"Multiplication: {x} * {y} = {result_multiply}")

    # 4. Division (Normal Case)
    result_divide_ok = divide(x, y)
    print(f"Division (OK): {x} / {y} = {result_divide_ok}")

    # 5. Division (Division by Zero Check)
    result_divide_error = divide(x, z)
    print(f"Division (Error): {x} / {z} = {result_divide_error}")

# Standard Python idiom to call the main function when the script is executed
if __name__ == "__main__":
    main()
